src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,183 @@
+    const Addr pc_addr = _speculativePC.instAddr();
+    const Addr req_vaddr = (pc_addr & BaseCPU::PCMask) + fetchOffset;
+
+    const RequestPtr fetch_req = std::make_shared<Request>();
+
+    // Set Ids for tracing requests through the system
+    fetch_req->taskId(_cpuPtr->taskId());
+    fetch_req->setContext(_committedState->contextId()); // May need to store
+                                                        // speculative variant
+                                                        // if we allow out-of-
+                                                        // order execution
+                                                        // between context
+                                                        // changes?
+
+    fetch_req->setVirt(0, req_vaddr, sizeof(TheISA::MachInst),
+                       Request::INST_FETCH, _cpuPtr->instMasterID(), pc_addr);
+
+    if (!_cpuPtr->requestInstAddrTranslation(fetch_req,
+        [this](Fault f, const RequestPtr& r) {onPCTranslated(f, r);})) {
+        panic("The CPU rejected my fetch translate request and I haven't been "
+            "programmed to know how to continue!!!");
+        // TODO reschedule a fetch for the future?
+    }
+}
+
+void
+SDCPUThread::onPCTranslated(Fault fault, const RequestPtr& req)
+{
+    if (fault != NoFault) {
+        // TODO handle fault (Store fault for commit stage, then stop further
+        // fetching)
+    }
+
+    // No changes to the request before asking the CPU to handle it if there is
+    // not a fault. Let the CPU generate the packet.
+    if (!_cpuPtr->requestInstruction(req,
+        [this](Fault f, TheISA::MachInst i) {
+            onInstructionFetched(f, i);
+        })) {
+        panic("The CPU rejected my fetch request and I haven't been programmed"
+            " to know how to continue!!!");
+    }
+}
+
+void
+SDCPUThread::onInstructionFetched(Fault fault, const TheISA::MachInst inst)
+{
+    if (fault != NoFault) {
+        // TODO handle fault
+    }
+
+    // Capure the PC state at the point of instruction retrieval.
+    TheISA::PCState pc_state = _speculativePC;
+    Addr fetched_addr = (pc_state.instAddr() & BaseCPU::PCMask) + fetchOffset;
+
+    decoder.moreBytes(pc_state, fetched_addr, inst);
+
+    StaticInstPtr decode_result = decoder.decode(pc_state);
+    // Will need to copy the object out if we want to use shared_ptr, since
+    // otherwise we risk a double deletion if it's on the same memory location.
+
+    if (decode_result) { // If a complete instruction was decoded
+        // TODO issue the instruction
+        fetchOffset = 0;
+    } else { // If we still need to fetch more MachInsts.
+        fetchOffset += sizeof(TheISA::MachInst);
+    }
+
+
+    if (true/* Replace with condition*/) {
+        // If we are still running and know where to fetch, we should
+        // immediately send the necessary requests to the CPU to fetch the next
+        // instruction
+        onBeginFetch();
```

Comment:
Hmm... I think we might want to rethink the fetch part of the pipeline. But let's get something working first.